DurableExecutorServiceTest.testDestroyCleansAllContainers

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3748/testReport/junit/com.hazelcast.durableexecutor/DurableExecutorServiceTest/testDestroyCleansAllContainers/

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/4935/testReport/junit/com.hazelcast.durableexecutor/DurableExecutorServiceTest/testDestroyCleansAllContainers/

```
java.lang.AssertionError: 
Some partitions has non-null containers for executor 30a426cb-0f50-44b3-9282-c63759534c5e:
Partition 99

	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.durableexecutor.DurableExecutorServiceTest$1.run(DurableExecutorServiceTest.java:136)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1345)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1362)
	at com.hazelcast.durableexecutor.DurableExecutorServiceTest.testDestroyCleansAllContainers(DurableExecutorServiceTest.java:122)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

```
Hiccups measured while running test 'testDestroyCleansAllContainers(com.hazelcast.durableexecutor.DurableExecutorServiceTest):'
07:15:05, accumulated pauses: 681 ms, max pause: 163 ms, pauses over 1000 ms: 0
07:15:10, accumulated pauses: 438 ms, max pause: 11 ms, pauses over 1000 ms: 0
07:15:15, accumulated pauses: 386 ms, max pause: 9 ms, pauses over 1000 ms: 0
07:15:20, accumulated pauses: 267 ms, max pause: 9 ms, pauses over 1000 ms: 0
07:15:25, accumulated pauses: 237 ms, max pause: 8 ms, pauses over 1000 ms: 0
07:15:30, accumulated pauses: 356 ms, max pause: 8 ms, pauses over 1000 ms: 0
07:15:35, accumulated pauses: 300 ms, max pause: 9 ms, pauses over 1000 ms: 0
07:15:40, accumulated pauses: 374 ms, max pause: 318 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/main/java/com/hazelcast/durableexecutor/impl/DurableExecutorPartitionContainer.java
+++ b/hazelcast/src/main/java/com/hazelcast/durableexecutor/impl/DurableExecutorPartitionContainer.java
@@ -18,19 +18,21 @@ package com.hazelcast.durableexecutor.impl;
 
 import com.hazelcast.config.DurableExecutorConfig;
 import com.hazelcast.durableexecutor.impl.operations.ReplicationOperation;
-import com.hazelcast.spi.impl.operationservice.Operation;
 import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.impl.operationservice.Operation;
 
 import java.util.HashMap;
 import java.util.Iterator;
-import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 
 public class DurableExecutorPartitionContainer {
 
     private final int partitionId;
     private final NodeEngineImpl nodeEngine;
 
-    private final Map<String, DurableExecutorContainer> executorContainerMap = new HashMap<>();
+    private final ConcurrentMap<String, DurableExecutorContainer> executorContainerMap
+            = new ConcurrentHashMap<>();
 
     public DurableExecutorPartitionContainer(NodeEngineImpl nodeEngine, int partitionId) {
         this.nodeEngine = nodeEngine;

--- a/hazelcast/src/test/java/com/hazelcast/durableexecutor/DurableExecutorServiceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/durableexecutor/DurableExecutorServiceTest.java
@@ -29,7 +29,6 @@ import com.hazelcast.durableexecutor.impl.DurableExecutorContainer;
 import com.hazelcast.executor.ExecutorServiceTestSupport;
 import com.hazelcast.partition.PartitionAware;
 import com.hazelcast.spi.properties.ClusterProperty;
-import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.ParallelJVMTest;
@@ -107,11 +106,11 @@ public class DurableExecutorServiceTest extends ExecutorServiceTestSupport {
     public void testDestroyCleansAllContainers() throws Exception {
         final String executorName = randomMapName();
         TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(NODE_COUNT);
-        final HazelcastInstance[] instances = factory.newInstances();
+        final HazelcastInstance[] instances = factory.newInstances(smallInstanceConfig());
 
         DurableExecutorService service = instances[0].getDurableExecutorService(executorName);
 
-        List<Future<String>> futures = new ArrayList<Future<String>>(TASK_COUNT);
+        List<Future<String>> futures = new ArrayList<>(TASK_COUNT);
         for (int i = 0; i < TASK_COUNT; i++) {
             futures.add(service.submit(new DummyCallable()));
         }
@@ -119,22 +118,25 @@ public class DurableExecutorServiceTest extends ExecutorServiceTestSupport {
             future.get();
         }
         service.destroy();
-        assertTrueEventually(new AssertTask() {
-            @Override
-            public void run() {
-                for (int i = 0; i < NODE_COUNT; i++) {
-                    DistributedDurableExecutorService internalService = getNodeEngineImpl(instances[i]).getService(DistributedDurableExecutorService.SERVICE_NAME);
-                    boolean allEmpty = true;
-                    StringBuilder failMessage = new StringBuilder();
-                    for (int partitionId = 0; partitionId < getNode(instances[i]).getProperties().getInteger(PARTITION_COUNT); partitionId++) {
-                        DurableExecutorContainer container = getDurableExecutorContainer(internalService, partitionId, executorName);
-                        if (container != null) {
-                            failMessage.append(String.format("Partition %d\n", partitionId));
-                            allEmpty = false;
-                        }
+        assertTrueEventually(() -> {
+            for (int i = 0; i < NODE_COUNT; i++) {
+                DistributedDurableExecutorService internalService = getNodeEngineImpl(instances[i])
+                        .getService(DistributedDurableExecutorService.SERVICE_NAME);
+                boolean allEmpty = true;
+                StringBuilder failMessage = new StringBuilder();
+                for (int partitionId = 0;
+                     partitionId < getNode(instances[i]).getProperties().getInteger(PARTITION_COUNT);
+                     partitionId++) {
+                    DurableExecutorContainer container = getDurableExecutorContainer(internalService, partitionId, executorName);
+                    if (container != null) {
+                        failMessage.append(String.format("Partition %d owned by %s\n", partitionId,
+                                getPartitionService(instances[0]).getPartition(partitionId).getOwnerOrNull()));
+                        allEmpty = false;
                     }
-                    assertTrue(String.format("Some partitions has non-null containers for executor %s:\n%s", executorName, failMessage.toString()), allEmpty);
                 }
+                assertTrue(String.format("Some partitions have non-null containers for executor %s:\n%s",
+                        executorName,
+                        failMessage.toString()), allEmpty);
             }
         }, 30);
     }

